pkg/server/api/localauthority/v1/service.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	if s.ca.GetNextX509CASlot().IsEmpty() {
+		return nil, api.MakeErr(log, codes.Internal, "no prepared authority found", nil)
+	}
+
+	// Move next into current and reset next to clean CA
+	s.ca.RotateX509CA()
+
+	current := s.ca.GetCurrentX509CASlot()
+	state := &localauthorityv1.AuthorityState{
+		AuthorityId: current.AuthorityID(),
+		Status:      localauthorityv1.AuthorityState_ACTIVE,
+	}
+	rpccontext.AuditRPC(ctx)
+
+	return &localauthorityv1.ActivateX509AuthorityResponse{
+		ActivatedAuthority: state,
+	}, nil
+}
+
+func (s *Service) TaintX509Authority(ctx context.Context, req *localauthorityv1.TaintX509AuthorityRequest) (*localauthorityv1.TaintX509AuthorityResponse, error) {
+	rpccontext.AddRPCAuditFields(ctx, buildAuditLogFields(req.AuthorityId))
+	log := rpccontext.Logger(ctx)
+
+	authorityID, publicKey, err := s.getX509PublicKey(ctx, req.AuthorityId)
+	if err != nil {
+		if req.AuthorityId != "" {
+			log = log.WithField(telemetry.LocalAuthorityID, req.AuthorityId)
+		}
+		return nil, api.MakeErr(log, codes.InvalidArgument, "invalid authority ID", err)
+	}
+	log = log.WithField(telemetry.LocalAuthorityID, authorityID)
+
+	if err := s.ds.TaintX509CA(ctx, s.td.IDString(), publicKey); err != nil {
+		return nil, api.MakeErr(log, codes.Internal, "failed to taint X.509 authority", err)
+	}
+
+	state := &localauthorityv1.AuthorityState{
+		AuthorityId: authorityID,
+		Status:      localauthorityv1.AuthorityState_OLD,
+	}
+
+	rpccontext.AuditRPC(ctx)
+	log.Info("X.509 authority tainted successfully")
+
+	return &localauthorityv1.TaintX509AuthorityResponse{
+		TaintedAuthority: state,
+	}, nil
+}
+
+func (s *Service) RevokeX509Authority(ctx context.Context, req *localauthorityv1.RevokeX509AuthorityRequest) (*localauthorityv1.RevokeX509AuthorityResponse, error) {
+	rpccontext.AddRPCAuditFields(ctx, buildAuditLogFields(req.AuthorityId))
+	log := rpccontext.Logger(ctx)
+
+	authorityID, publicKey, err := s.getX509PublicKey(ctx, req.AuthorityId)
+	if err != nil {
+		if req.AuthorityId != "" {
+			log = log.WithField(telemetry.LocalAuthorityID, req.AuthorityId)
+		}
+		return nil, api.MakeErr(log, codes.InvalidArgument, "invalid authority ID", err)
+	}
+
+	log = log.WithField(telemetry.LocalAuthorityID, authorityID)
+	if err := s.ds.RevokeX509CA(ctx, s.td.IDString(), publicKey); err != nil {
+		return nil, api.MakeErr(log, codes.Internal, "failed to revoke X.509 authority", err)
+	}
+
+	state := &localauthorityv1.AuthorityState{
+		AuthorityId: authorityID,
+		Status:      localauthorityv1.AuthorityState_OLD,
+	}
+
+	rpccontext.AuditRPC(ctx)
+	log.Info("X.509 authority revoked successfully")
+
+	return &localauthorityv1.RevokeX509AuthorityResponse{
+		RevokedAuthority: state,
+	}, nil
+}
+
+// getX509PublicKey validates provided authority ID, and return OLD associated publick key, in case of authority ID is not provided, use the current OLD authority
```

Comment:
```suggestion
// getX509PublicKey validates provided authority ID, and return OLD associated public key, in case of authority ID is not provided, use the current OLD authority
```

Relying on empty authority id to mean the old authority feels kind of dangerous (related to above comment about missing authority id in req)

Reply:
so do you suggest here to make authorityID required, but it must be the old on?